operator/validator/metadata/syncer.go

```diff
@@ -0,0 +1,366 @@
+}
+
+// nextBatch returns non-liquidated shares from DB that are most deserving of an update, it relies on share.Metadata.lastUpdated to be updated in order to keep iterating forward.
+func (s *Syncer) nextBatch(_ context.Context, subnetsBuf *big.Int) []*ssvtypes.SSVShare {
+	// TODO: use context, return if it's done
+	ownSubnets := s.selfSubnets(subnetsBuf)
+
+	var staleShares, newShares []*ssvtypes.SSVShare
+	s.shareStorage.Range(nil, func(share *ssvtypes.SSVShare) bool {
+		if share.Liquidated {
+			return true
+		}
+
+		networkcommons.SetCommitteeSubnet(subnetsBuf, share.CommitteeID())
+		subnet := subnetsBuf.Uint64()
+		if ownSubnets[subnet] == 0 {
+			return true
+		}
+
+		shareMetadataStaleAfter := s.syncInterval
+		// also check last metadata update to differentiate new shares from stale ones
+		if !share.HasBeaconMetadata() && share.MetadataLastUpdated().IsZero() {
+			newShares = append(newShares, share)
+		} else if time.Since(share.MetadataLastUpdated()) > shareMetadataStaleAfter {
+			staleShares = append(staleShares, share)
+		}
+		return len(newShares) < batchSize
+	})
+
+	// Combine validators up to batchSize, prioritizing the new ones.
+	shares := newShares
+	if remainder := batchSize - len(shares); remainder > 0 {
+		end := remainder
+		if end > len(staleShares) {
+			end = len(staleShares)
+		}
+		shares = append(shares, staleShares[:end]...)
+	}
+
+	for _, share := range shares {
+		share.SetMetadataLastUpdated(time.Now())
+	}
+	return shares
+}
+
+// TODO: Create a wrapper for share storage that contains all common methods like AllActiveIndices and use the wrapper.
+func (s *Syncer) allActiveIndices(_ context.Context, epoch phase0.Epoch) []phase0.ValidatorIndex {
+	var indices []phase0.ValidatorIndex
+
+	// TODO: use context, return if it's done
+	s.shareStorage.Range(nil, func(share *ssvtypes.SSVShare) bool {
+		if share.IsParticipating(epoch) {
+			indices = append(indices, share.BeaconMetadata.Index)
+		}
+		return true
+	})
+
+	return indices
+}
+
+func (s *Syncer) sleep(ctx context.Context, d time.Duration) (slept bool) {
+	// TODO: use time.After when Go is updated to 1.23
+	timer := time.NewTimer(d)
+	defer timer.Stop()
+
+	select {
+	case <-ctx.Done():
+		return false
+	case <-timer.C:
+		return true
+	}
+}
+
+// selfSubnets calculates the operator's subnets by adding up the fixed subnets and the active committees
+// it recvs big int buffer for memory reusing, if is nil it will allocate new
+func (s *Syncer) selfSubnets(buf *big.Int) records.Subnets {
+	// Start off with a copy of the fixed subnets (e.g., exporter subscribed to all subnets).
+	localBuf := buf
+	if localBuf == nil {
+		localBuf = new(big.Int)
+	}
+
+	myValidators := s.validatorStore.SelfValidators()
+	mySubnets := make(records.Subnets, networkcommons.SubnetsCount)
+	for _, v := range myValidators {
+		networkcommons.SetCommitteeSubnet(localBuf, v.CommitteeID())
+		mySubnets[localBuf.Uint64()] = 1
+	}
+
+	return mySubnets
+}
```

Comment:
this needs to also include FixedSubnets, you might want to pass the config var to here as well..

Reply:
done